File: Tests/VIPSImageTestCase.m
//
//  VIPSImageTestCase.m
//  VIPSKitTests
//
//  Base test case class with shared setup and helper methods
//

#import "VIPSImageTestCase.h"

static BOOL sVIPSInitialized = NO;

@implementation VIPSImageTestCase

- (void)setUp {
    [super setUp];
    if (!sVIPSInitialized) {
        NSError *error = nil;
        BOOL success = [VIPSImage initializeWithError:&error];
        XCTAssertTrue(success, @"Failed to initialize VIPSKit: %@", error);
        sVIPSInitialized = success;
    }
    XCTAssertTrue(sVIPSInitialized, @"VIPSKit not initialized");
}

#pragma mark - Helper Methods

- (NSString *)pathForTestResource:(NSString *)filename {
    // Try bundle resource path first (for running in Xcode)
    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
    NSString *path = [bundle pathForResource:[filename stringByDeletingPathExtension]
                                      ofType:[filename pathExtension]];
    if (path) {
        return path;
    }

    // Fallback to hardcoded path for command-line testing
    NSString *testResourcesPath = @"/Users/TiM/Developer/VIPSKit/Tests/TestResources";
    path = [testResourcesPath stringByAppendingPathComponent:filename];

    if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
        return path;
    }

    return nil;
}

- (VIPSImage *)createTestImageWithWidth:(NSInteger)width height:(NSInteger)height {
    return [self createTestImageWithWidth:width height:height bands:3];
}

- (VIPSImage *)createTestImageWithWidth:(NSInteger)width height:(NSInteger)height bands:(NSInteger)bands {
    NSMutableData *buffer = [NSMutableData dataWithLength:width * height * bands];

    // Fill with gradient
    unsigned char *bytes = (unsigned char *)buffer.mutableBytes;
    for (NSInteger y = 0; y < height; y++) {
        for (NSInteger x = 0; x < width; x++) {
            NSInteger idx = (y * width + x) * bands;
            bytes[idx + 0] = (unsigned char)(x * 255 / width);     // R
            if (bands >= 2) bytes[idx + 1] = (unsigned char)(y * 255 / height);    // G
            if (bands >= 3) bytes[idx + 2] = 128;                  // B
            if (bands >= 4) bytes[idx + 3] = 200;                  // A
        }
    }

    NSError *error = nil;
    VIPSImage *image = [VIPSImage imageWithBuffer:buffer.bytes width:width height:height bands:bands error:&error];
    XCTAssertNotNil(image, @"Failed to create test image: %@", error);
    return image;
}

- (VIPSImage *)createSolidColorImageWithWidth:(NSInteger)width
                                       height:(NSInteger)height
                                          red:(uint8_t)r
                                        green:(uint8_t)g
                                         blue:(uint8_t)b {
    NSMutableData *buffer = [NSMutableData dataWithLength:width * height * 3];
    unsigned char *bytes = (unsigned char *)buffer.mutableBytes;

    for (NSInteger i = 0; i < width * height; i++) {
        bytes[i * 3 + 0] = r;
        bytes[i * 3 + 1] = g;
        bytes[i * 3 + 2] = b;
    }

    NSError *error = nil;
    VIPSImage *image = [VIPSImage imageWithBuffer:buffer.bytes width:width height:height bands:3 error:&error];
    XCTAssertNotNil(image, @"Failed to create solid color image: %@", error);
    return image;
}

- (VIPSImage *)createSolidColorImageWithWidth:(NSInteger)width
                                       height:(NSInteger)height
                                          red:(uint8_t)r
                                        green:(uint8_t)g
                                         blue:(uint8_t)b
                                        alpha:(uint8_t)a {
    NSMutableData *buffer = [NSMutableData dataWithLength:width * height * 4];
    unsigned char *bytes = (unsigned char *)buffer.mutableBytes;

    for (NSInteger i = 0; i < width * height; i++) {
        bytes[i * 4 + 0] = r;
        bytes[i * 4 + 1] = g;
        bytes[i * 4 + 2] = b;
        bytes[i * 4 + 3] = a;
    }

    NSError *error = nil;
    VIPSImage *image = [VIPSImage imageWithBuffer:buffer.bytes width:width height:height bands:4 error:&error];
    XCTAssertNotNil(image, @"Failed to create solid color RGBA image: %@", error);
    return image;
}

- (VIPSImage *)createImageWithMarginsWidth:(NSInteger)width
                                    height:(NSInteger)height
                                marginSize:(NSInteger)margin
                             contentColorR:(uint8_t)contentR
                             contentColorG:(uint8_t)contentG
                             contentColorB:(uint8_t)contentB
                          backgroundColorR:(uint8_t)bgR
                          backgroundColorG:(uint8_t)bgG
                          backgroundColorB:(uint8_t)bgB {
    NSMutableData *buffer = [NSMutableData dataWithLength:width * height * 3];
    unsigned char *bytes = (unsigned char *)buffer.mutableBytes;

    for (NSInteger y = 0; y < height; y++) {
        for (NSInteger x = 0; x < width; x++) {
            NSInteger idx = (y * width + x) * 3;

            // Check if this pixel is in the margin area
            BOOL inMargin = (x < margin || x >= width - margin ||
                            y < margin || y >= height - margin);

            if (inMargin) {
                bytes[idx + 0] = bgR;
                bytes[idx + 1] = bgG;
                bytes[idx + 2] = bgB;
            } else {
                bytes[idx + 0] = contentR;
                bytes[idx + 1] = contentG;
                bytes[idx + 2] = contentB;
            }
        }
    }

    NSError *error = nil;
    VIPSImage *image = [VIPSImage imageWithBuffer:buffer.bytes width:width height:height bands:3 error:&error];
    XCTAssertNotNil(image, @"Failed to create image with margins: %@", error);
    return image;
}

- (VIPSImage *)createHorizontalGradientWidth:(NSInteger)width
                                      height:(NSInteger)height
                                 startColorR:(uint8_t)startR
                                 startColorG:(uint8_t)startG
                                 startColorB:(uint8_t)startB
                                   endColorR:(uint8_t)endR
                                   endColorG:(uint8_t)endG
                                   endColorB:(uint8_t)endB {
    NSMutableData *buffer = [NSMutableData dataWithLength:width * height * 3];
    unsigned char *bytes = (unsigned char *)buffer.mutableBytes;

    for (NSInteger y = 0; y < height; y++) {
        for (NSInteger x = 0; x < width; x++) {
            NSInteger idx = (y * width + x) * 3;
            double t = (double)x / (double)(width - 1);

            bytes[idx + 0] = (uint8_t)(startR + t * (endR - startR));
            bytes[idx + 1] = (uint8_t)(startG + t * (endG - startG));
            bytes[idx + 2] = (uint8_t)(startB + t * (endB - startB));
        }
    }

    NSError *error = nil;
    VIPSImage *image = [VIPSImage imageWithBuffer:buffer.bytes width:width height:height bands:3 error:&error];
    XCTAssertNotNil(image, @"Failed to create gradient image: %@", error);
    return image;
}

@end

File: Tests/VIPSImageTestCase.swift
import XCTest
@testable import VIPSKit

class VIPSImageTestCase: XCTestCase {

    private static var initialized = false

    override func setUp() {
        super.setUp()
        if !Self.initialized {
            XCTAssertNoThrow(try VIPSImage.initialize())
            Self.initialized = true
        }
    }

    // MARK: - Helpers

    func pathForTestResource(_ filename: String) -> String? {
        let bundle = Bundle(for: type(of: self))
        let name = (filename as NSString).deletingPathExtension
        let ext = (filename as NSString).pathExtension
        if let path = bundle.path(forResource: name, ofType: ext) {
            return path
        }
        let fallback = "/Users/TiM/Developer/VIPSKit/Tests/TestResources/\(filename)"
        if FileManager.default.fileExists(atPath: fallback) {
            return fallback
        }
        return nil
    }

    func createTestImage(width: Int, height: Int, bands: Int = 3) -> VIPSImage {
        var buffer = [UInt8](repeating: 0, count: width * height * bands)
        for y in 0..<height {
            for x in 0..<width {
                let idx = (y * width + x) * bands
                buffer[idx] = UInt8(x * 255 / max(width - 1, 1))
                if bands >= 2 { buffer[idx + 1] = UInt8(y * 255 / max(height - 1, 1)) }
                if bands >= 3 { buffer[idx + 2] = 128 }
                if bands >= 4 { buffer[idx + 3] = 200 }
            }
        }
        return try! VIPSImage(buffer: &buffer, width: width, height: height, bands: bands)
    }

    func createSolidColorImage(width: Int, height: Int, r: UInt8, g: UInt8, b: UInt8) -> VIPSImage {
        var buffer = [UInt8](repeating: 0, count: width * height * 3)
        for i in 0..<(width * height) {
            buffer[i * 3] = r; buffer[i * 3 + 1] = g; buffer[i * 3 + 2] = b
        }
        return try! VIPSImage(buffer: &buffer, width: width, height: height, bands: 3)
    }

    func createSolidColorImage(width: Int, height: Int, r: UInt8, g: UInt8, b: UInt8, a: UInt8) -> VIPSImage {
        var buffer = [UInt8](repeating: 0, count: width * height * 4)
        for i in 0..<(width * height) {
            buffer[i * 4] = r; buffer[i * 4 + 1] = g; buffer[i * 4 + 2] = b; buffer[i * 4 + 3] = a
        }
        return try! VIPSImage(buffer: &buffer, width: width, height: height, bands: 4)
    }

    func createImageWithMargins(width: Int, height: Int, margin: Int,
                                contentR: UInt8, contentG: UInt8, contentB: UInt8,
                                bgR: UInt8, bgG: UInt8, bgB: UInt8) -> VIPSImage {
        var buffer = [UInt8](repeating: 0, count: width * height * 3)
        for y in 0..<height {
            for x in 0..<width {
                let idx = (y * width + x) * 3
                let inMargin = x < margin || x >= width - margin || y < margin || y >= height - margin
                if inMargin {
                    buffer[idx] = bgR; buffer[idx + 1] = bgG; buffer[idx + 2] = bgB
                } else {
                    buffer[idx] = contentR; buffer[idx + 1] = contentG; buffer[idx + 2] = contentB
                }
            }
        }
        return try! VIPSImage(buffer: &buffer, width: width, height: height, bands: 3)
    }

    func createHorizontalGradient(width: Int, height: Int,
                                  startR: UInt8, startG: UInt8, startB: UInt8,
                                  endR: UInt8, endG: UInt8, endB: UInt8) -> VIPSImage {
        var buffer = [UInt8](repeating: 0, count: width * height * 3)
        for y in 0..<height {
            for x in 0..<width {
                let idx = (y * width + x) * 3
                let t = Double(x) / Double(max(width - 1, 1))
                buffer[idx] = UInt8(Double(startR) + t * Double(Int(endR) - Int(startR)))
                buffer[idx + 1] = UInt8(Double(startG) + t * Double(Int(endG) - Int(startG)))
                buffer[idx + 2] = UInt8(Double(startB) + t * Double(Int(endB) - Int(startB)))
            }
        }
        return try! VIPSImage(buffer: &buffer, width: width, height: height, bands: 3)
    }
}
